### PR TITLE
Add RazorDiagnostic type

### PR DESCRIPTION
**Pull request by rynowak**
This is a replacement for RazorError, is conceptually equivalent with
Diagnostic from Roslyn.

The next PR will start exposing this through our public API rather than
the legacy type.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorDiagnostic.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        public override bool Equals(RazorDiagnostic obj)
+        {
+            var other = obj as DefaultRazorDiagnostic;
+            if (other == null)
+            {
+                return false;
+            }
+
+            if (!_descriptor.Equals(other._descriptor))
+                {
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Evolution/DefaultRazorDiagnostic.cs]: formatting

File: src/Microsoft.AspNetCore.Razor.Evolution/RazorDiagnosticDescriptor.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Razor.Evolution
+{
+    [DebuggerDisplay("{Severity - {Id}: {GetMessageFormat())")]
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Evolution/RazorDiagnosticDescriptor.cs]: Braces look unbalanced

File: src/Microsoft.AspNetCore.Razor.Evolution/Resources.resx [hunk 3]
```diff
@@ -117,6 +117,9 @@
   <resheader name="writer">
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
+  <data name="Argument_Cannot_Be_Null_Or_Empty" xml:space="preserve">
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Evolution/Resources.resx]: `ArgumentCannotBeNullOrEmpty`

File: src/Microsoft.AspNetCore.Razor.Evolution/RazorDiagnosticDescriptor.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            {
+                return false;
+            }
+
+            return string.Equals(Id, other.Id, StringComparison.Ordinal);
+        }
+
+        public override int GetHashCode()
+        {
+            return Id.GetHashCode();
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Razor.Evolution/RazorDiagnosticDescriptor.cs]: `StringComparer.Ordinal.GetHashCode(Id);`

**Comment by rynowak** [comment]
e549270b4a5040f58dbb6fa98018fb0dbaff3a4e

**Pull request closed without merging**